first_task.scala

```diff
@@ -0,0 +1,260 @@
+      .option("multiLine", true)
+      .option("header", true)
+      .schema(Encoders.product[Answer].schema)
+      .option("escape", "\"")
+      .csv("/FileStore/tables/Answers.csv")
+      .na.drop("all")
+      .as[Answer]
+
+answersDs.cache()
+
+// COMMAND ----------
+
+case class Tag(id: Long,
+               tag: String)
+
+val tagsDs = spark.read
+      .option("multiLine", true)
+      .option("header", true)
+      .schema(Encoders.product[Tag].schema)
+      .option("escape", "\"")
+      .csv("/FileStore/tables/Tags.csv")
+      .na.drop("all")
+      .as[Tag]
+
+tagsDs.cache()
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Collect daily statistics (number of unique users posted, number of questions and answers) into a separate table and export to a Parquet file.
+
+// COMMAND ----------
+
+val questionNaFreeDs = questionsDs
+      .withColumn("creationDate", col("creationDate").cast(DateType))
+      .na.drop(Seq("creationDate"))
+val answersNaFreeDs = answersDs
+  .withColumn("creationDate", col("creationDate").cast(DateType))
+  .na.drop(Seq("creationDate"))
+
+val questionCols = Seq(
+  lit(null).alias("answerId"),
+  col("ownerUserId"),
+  col("creationDate").cast(DateType),
+  col("id").alias("questionId")
+)
+val answerCols = Seq(
+  col("id").alias("answerId"),
+  col("ownerUserId"),
+  col("creationDate").cast(DateType),
+  lit(null).alias("questionId")
+)
+val unionDf = questionNaFreeDs.select(questionCols: _*).unionAll(answersNaFreeDs.select(answerCols: _ *))
+val dailyStatisticsDf = unionDf
+  .groupBy("creationDate")
+  .agg(
+  countDistinct("ownerUserId").alias("unique_users"),
+  count("questionId").alias("number_of_questions"),
+  count("answerId").alias("number_of_answers")
+)
+
+dbutils.fs.rm("/FileStore/tables/daily.parquet", true) 
+dailyStatisticsDf.write.parquet("/FileStore/tables/daily.parquet")
+dailyStatisticsDf.show()
+
+// COMMAND ----------
+
+def countMeanAndMedianOfAnswerScores(dataset: Dataset[Answer], predicate: Answer => Boolean): (Double, Double) = {
+    val mean = dataset
+      .filter(predicate)
+      .select(avg("score"))
+      .take(1)(0)(0).toString().toDouble
+
+    val median = dataset.filter(predicate).stat.approxQuantile("score", Array(0.5), 0.1)(0)
+
+    (mean, median)
+}
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC What is the average and median score of the answers containing references (links) vs. those which don't have any refs(links)?
+
+// COMMAND ----------
+
+val containsLink = (a: Answer) => !Jsoup.parse(a.body).select("a[href]").isEmpty
+println(countMeanAndMedianOfAnswerScores(answersDs,containsLink(_)))
+println(countMeanAndMedianOfAnswerScores(answersDs, !containsLink(_)))
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC What is the average and median score of the answers containing code snippets against those who don't have any code snippets?
+
+// COMMAND ----------
+
+val containsLink = (a: Answer) => !Jsoup.parse(a.body).select("code").isEmpty
```

Comment:
containsCode?

Reply:
And if this is containsCode, this probably won't answer my question on this dataset.
Not everything wrapped up into a <code> tag is a code snippet. Please compare this:
![image](https://user-images.githubusercontent.com/10655440/63789182-87ad2d80-c8ff-11e9-88c8-3f7c17cea684.png)
and that:
![image](https://user-images.githubusercontent.com/10655440/63789390-f5595980-c8ff-11e9-96d2-ab3ad266d45b.png)
in the second case it's a code snippet, in the first case it's not.